File: src/dss.ts
import got from 'got';

const headers = {'user-agent': 'Mozilla/5.0 (X11; Ubuntu; Linux x86_64; rv:87.0) Gecko/20100101 Firefox/87.0'};
const LOCALIZED_NO_RESULT_FOR_SEARCH_TERM = 'Nessun risultato trovato per i termini di ricerca';

export class Dss
{
    public static async count_event(query: string): Promise<number>
    {
        const search_key = `"${query.trim().replace(' ', '+')}"`;

        try
        {
            const result = await got(`https://www.google.it/search?q=${search_key}&gws_rd=cr,ssl&ei=odOnVsHMBcKke7jerogD`, {headers});

            if(!result.body.includes('<div id="result-stats">'))
                return 0;

            const matched = result.body.match(/<div id="result\-stats">([a-zA-Z 0-9\.]*)<nobr>/);
            const no_result = result.body.indexOf(LOCALIZED_NO_RESULT_FOR_SEARCH_TERM);

            if(no_result !== -1)
                return 0;
            
            if(!matched.length)
                return 0;

            const fields = matched[1].split(' ');
            const value = fields[fields.length > 2 ? 1 : 0].replace(/\./g,'');
            return parseInt(value, 10);
        }
        catch(error)
        {
            console.error(error);
            return 0;
        }
    }

    public static async count_events(head: string, keys: string[]): Promise<{keys: string[], values: number[]}>
    {
        const values = await Promise.all(keys.map(async (query) => {
            return await this.count_event(`${head} ${query}`);
        }));

        return {keys, values};
    }
}

File: src/dss/dss.go
package dss

import (
	"fmt"
	"io/ioutil"
	"net/http"
	"net/url"
	"regexp"
	"strconv"
	"strings"
	"sync"
	"time"
)

// Constants to be used in the search
const (
	UserAgent                      = "Mozilla/5.0 (X11; Ubuntu; Linux x86_64; rv:87.0) Gecko/20100101 Firefox/87.0"
	LocalizedNoResultForSearchTerm = "Nessun risultato trovato per i termini di ricerca"
	ResultStatsDivIDPattern        = `<div id="result\-stats">([a-zA-Z 0-9\.]*)<nobr>`
)

// Dss is the class containing methods for counting events
type Dss struct{}

// CountEvent fetches the Google search result for a query and returns the count of results
func (d *Dss) CountEvent(query string, after *time.Time) (int, error) {

	if after != nil {
		query = fmt.Sprintf("%s after:%s", query, after.Format("2006/01/02"))
	}

	searchKey := url.QueryEscape(`"` + strings.ReplaceAll(strings.TrimSpace(query), " ", "+") + `"`)
	url := fmt.Sprintf("https://www.google.it/search?q=%s", searchKey)

	// Create HTTP request
	client := &http.Client{
		Timeout: 30 * time.Second,
	}
	req, err := http.NewRequest("GET", url, nil)
	if err != nil {
		return 0, err
	}
	req.Header.Set("user-agent", UserAgent)
	req.Close = true

	// Execute the request
	resp, err := client.Do(req)
	if err != nil {
		return 0, err
	}
	defer resp.Body.Close()

	// Read response body
	body, err := ioutil.ReadAll(resp.Body)
	if err != nil {
		return 0, err
	}

	// Check if result stats div is present
	bodyStr := string(body)
	if !strings.Contains(bodyStr, `<div id="result-stats">`) {
		println(bodyStr)
		return 0, nil
	}

	// Match the result count using regex
	re := regexp.MustCompile(ResultStatsDivIDPattern)
	matched := re.FindStringSubmatch(bodyStr)
	if len(matched) == 0 {
		return 0, nil
	}

	// Check if no result localized string is present
	if strings.Contains(bodyStr, LocalizedNoResultForSearchTerm) {
		return 0, nil
	}

	// Parse the number from the matched text
	fields := strings.Fields(matched[1])

	pos := 0

	if len(fields) > 2 {
		pos = 1
	}

	value := strings.ReplaceAll(fields[pos], ".", "") // Remove dots from the number
	count, err := strconv.Atoi(value)
	if err != nil {
		println("Error parsing count", value)
		return 0, err
	}

	return count, nil
}

// CountEvents returns the number of results for each of the provided search keys combined with a head
func (d *Dss) CountEvents(head string, keys []string, after *time.Time) (map[string]int, error) {
	results := make(map[string]int)
	var wg sync.WaitGroup
	mu := &sync.Mutex{} // Mutex to prevent concurrent map writes

	// Goroutine for each query to make it concurrent
	for _, key := range keys {
		println("Start query for", head, " : ", key)
		wg.Add(1)
		go func(k string) {
			defer wg.Done()
			query := fmt.Sprintf("%s %s", head, k)
			println("Querying for", query)
			count, err := d.CountEvent(query, after)
			if err != nil {
				fmt.Printf("Error for query %s: %v\n", query, err)
				return
			}
			println("Count for", query, " : ", count)
			mu.Lock()
			results[k] = count
			mu.Unlock()
		}(key)
	}

	// Wait for all goroutines to finish
	wg.Wait()

	return results, nil
}
